src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala

```diff
@@ -819,26 +826,88 @@ class StoreQueue(implicit p: Parameters) extends XSModule
     is(s_wb) {
       when (io.mmioStout.fire || io.vecmmioStout.fire) {
         when (uncacheUop.exceptionVec(storeAccessFault)) {
-          uncacheState := s_idle
+          mmioState := s_idle
         }.otherwise {
-          uncacheState := s_wait
+          mmioState := s_wait
         }
       }
     }
     is(s_wait) {
       // A MMIO store can always move cmtPtrExt as it must be ROB head
       when(scommit > 0.U) {
-        uncacheState := s_idle // ready for next mmio
+        mmioState := s_idle // ready for next mmio
+      }
+    }
+  }
+
+  mmioReq.valid := mmioState === s_req
+  mmioReq.bits := DontCare
+  mmioReq.bits.cmd  := MemoryOpConstants.M_XWR
+  mmioReq.bits.addr := paddrModule.io.rdata(0) // data(deqPtr) -> rdata(0)
+  mmioReq.bits.vaddr:= vaddrModule.io.rdata(0)
+  mmioReq.bits.data := shiftDataToLow(paddrModule.io.rdata(0), dataModule.io.rdata(0).data)
+  mmioReq.bits.mask := shiftMaskToLow(paddrModule.io.rdata(0), dataModule.io.rdata(0).mask)
+  mmioReq.bits.atomic := atomic(GatedRegNext(rdataPtrExtNext(0)).value)
+  mmioReq.bits.nc := false.B
+  mmioReq.bits.id := rdataPtrExt(0).value
+
+  /**
+    * NC Store
+    * (1) req: when it has been commited, it can be sent to lower level.
+    * (2) resp: because SQ data forward is required, it can only be deq when ncResp is received
+    */
+  // TODO: CAN NOT deal with vector nc now!
+  val nc_idle :: nc_req :: nc_resp :: Nil = Enum(3)
```

Comment:
Since MMIO execution conditions are stricter than NC, `mmioState` are coupled with many signals of SQ. If `mmioState` and `ncState` are merged, it would make the code less intuitive and require additional checks to distinguish whether MMIO or NC for many coupled signals controlled by `mmioState`, increasing logical complexity. Keeping `ncState` separate only requires a 2-bit register and arbitration logic for `mmioReq` and `ncReq`. Therefore, I prefer not to merge `mmioState` and `ncState`.